plugins/line-highlight/prism-line-highlight.js

```diff
@@ -1,181 +1,211 @@
-
-		line.setAttribute('aria-hidden', 'true');
-		line.setAttribute('data-range', currentRange);
-		line.className = (classes || '') + ' line-highlight';
-
-		//if the line-numbers plugin is enabled, then there is no reason for this plugin to display the line numbers
-		if(hasLineNumbers && Prism.plugins.lineNumbers) {
-			var startNode = Prism.plugins.lineNumbers.getLine(pre, start);
-			var endNode = Prism.plugins.lineNumbers.getLine(pre, end);
-			
-			if (startNode) {
-				line.style.top = startNode.offsetTop + 'px';
+	function $$(expr, con) {
+		return Array.prototype.slice.call((con || document).querySelectorAll(expr));
+	}
+	
+	function hasClass(element, className) {
+	  className = " " + className + " ";
+	  return (" " + element.className + " ").replace(/[\n\t]/g, " ").indexOf(className) > -1
+	}
+	
+	// Some browsers round the line-height, others don't.
+	// We need to test for it to position the elements properly.
+	var isLineHeightRounded = (function() {
+		var res;
+		return function() {
+			if(typeof res === 'undefined') {
+				var d = document.createElement('div');
+				d.style.fontSize = '13px';
+				d.style.lineHeight = '1.5';
+				d.style.padding = 0;
+				d.style.border = 0;
+				d.innerHTML = '&nbsp;<br />&nbsp;';
+				document.body.appendChild(d);
+				// Browsers that round the line-height should have offsetHeight === 38
+				// The others should have 39.
+				res = d.offsetHeight === 38;
+				document.body.removeChild(d);
 			}
-			
-			if (endNode) {
-				line.style.height = (endNode.offsetTop - startNode.offsetTop) + endNode.offsetHeight + 'px';
+			return res;
+		}
+	}());
+	
+	function makeActions() {
+		var actions = [];
+	
+		return {
+			actions: actions,
+			apply: function(fn, context, args) {
+				actions.push(function() {
+					fn.apply(context, args);
+				});
+			},
+			set: function(obj, key, value) {
+				actions.push(function() {
+					obj[key] = value;
+				});
 			}
-		} else {
-			line.setAttribute('data-start', start);
-
-			if(end > start) {
-				line.setAttribute('data-end', end);
+		};
+	}
+	
+	function highlightLines(pre, lines, classes) {
+		lines = typeof lines === 'string' ? lines : pre.getAttribute('data-line');
+		
+		var ranges = lines.replace(/\s+/g, '').split(','),
+			offset = +pre.getAttribute('data-line-offset') || 0;
+	
+		var parseMethod = isLineHeightRounded() ? parseInt : parseFloat;
+		var lineHeight = parseMethod(getComputedStyle(pre).lineHeight);
+		var hasLineNumbers = hasClass(pre, 'line-numbers');
+		var action = makeActions();
+	
+		for (var i=0, currentRange; currentRange = ranges[i++];) {
+			var range = currentRange.split('-');
+	
+			var start = +range[0],
+				end = +range[1] || start;
+	
+			var line = pre.querySelector('.line-highlight[data-range="' + currentRange + '"]') || document.createElement('div');
+	
+			action.apply(line.setAttribute, line,['aria-hidden', 'true']);
+			action.apply(line.setAttribute, line,['data-range', currentRange]);
```

Comment:
I really don't know whether I like this style. It feels non-standard and kind of hacky.

That being said, I'm also unable to find a better solution because all of my approaches fail because of the scope of `var` variables. ES6 when?

Reply:
Which style?

Reply:
This function invocation style.

`action.apply(line.setAttribute, line,['aria-hidden', 'true']);` is exactly self-documenting IMO and it feels like it could be replaced by a simple lambda.

The lambda won't work because of the scoping of the captured variables, so we will most likely just go with this.

Reply:
These could be replaced by a simple function like

```
action.add(function(){
  line.setAttribute('aria-hidden', 'true')
})
```

But this is a bit slower as it creates functions and adds extra function calls.

Reply:
actually, there really isn't a performance gain.  I was pretending to be like CanJS's queues system, but didn't go all the way.  We can probably change it to the lambda style if you like that more.

Reply:
I don't think it can because the scope of the `line` variable is declared using `var`, so function scope outside of the loop.
This means that all of the created function will capture the same variable causing all of the created functions to have the same value for `line` when executed.

This is the exact problem which ES6 `let` and `const` solve if my understanding is correct.

Reply:
Also, you don't really need to be that concerned about the overhead of a single extra function call. It practically nonexistent compared to the performance bottle-nack it solves.

Reply:
Gotcha. Well, good to go then? I’m looking forward to snappy resizes.